common/utils.c

```diff
@@ -416,3 +416,106 @@ TDNFYesOrNo(
     return dwError;
 }
 
+uint32_t TDNFUriIsRemote(
+    const char* pszKeyUrl,
+    int *pnRemote
+)
+{
+    uint32_t dwError = 0;
+    const char *szRemotes[] = {"http://", "https://", "ftp://",
+                               "ftps://", NULL};
+    int i = 0;
+
+    if(!pnRemote || IsNullOrEmptyString(pszKeyUrl))
+    {
+      dwError = ERROR_TDNF_INVALID_PARAMETER;
+      BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    *pnRemote = 0;
+    for(i = 0; szRemotes[i]; i++) {
+        if (strncasecmp(pszKeyUrl, szRemotes[i], strlen(szRemotes[i])) == 0) {
+            *pnRemote = 1;
+            break;
+        }
+    }
+    if (szRemotes[i] == NULL && strncasecmp(pszKeyUrl, "file://", 7) != 0) {
+       dwError = ERROR_TDNF_KEYURL_INVALID;
+       BAIL_ON_TDNF_ERROR(dwError);
+    }
+cleanup:
+    return dwError;
+
+error:
+    goto cleanup;
+}
+
+uint32_t TDNFFileNameFromUri(
+    const char* pszKeyUrl,
+    char** ppszFile)
+{
+    uint32_t dwError = 0;
+    const char* pszPath = NULL;
+    char *pszFile = NULL;
+    size_t nOffset;
+    const char *szPrefixes[] = {"http://", "https://", "ftp://",
+                                "ftps://", "file://", NULL};
+    int i = 0;
+
+    if(IsNullOrEmptyString(pszKeyUrl) || !ppszFile)
+    {
+      dwError = ERROR_TDNF_INVALID_PARAMETER;
+      BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    for(i = 0; szPrefixes[i]; i++) {
+        if (strncasecmp(pszKeyUrl, szPrefixes[i], strlen(szPrefixes[i])) == 0) {
+            nOffset = strlen(szPrefixes[i]);
+            break;
+        }
+    }
+    if (szPrefixes[i] == NULL) {
+       dwError = ERROR_TDNF_KEYURL_INVALID;
+       BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pszPath = pszKeyUrl + nOffset;
+
+    if(!pszPath || *pszPath == '\0')
+    {
+        dwError = ERROR_TDNF_KEYURL_INVALID;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (strchr (pszPath, '#') != NULL)
+    {
+        dwError = ERROR_TDNF_KEYURL_INVALID;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if(*pszPath != '/')
+    {
+        //skip hostname in the uri.
+        pszPath = strchr (pszPath, '/');
+        if(pszPath == NULL)
+        {
+            dwError = ERROR_TDNF_KEYURL_INVALID;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
```

Comment:
This has some issues. Instead of `strchr` use `strrchr` to get the last occurrence of '/' and and then see if you > 1 chars in your remaining buf space (post striping '#' and '?').

Reply:
I don't follow - this is to skip over the hostname, if any. For example, if we started with "http://vmware.com/pathto/key.asc", `pszPath` would point to "vmware.com/pathto/key.asc". `strchr(pszPath, '/')` skips over the "vmware.com" so we get "/pasthto/key.asc", which is what we want.

OTOH, `strrchr()` would stop at "/key.asc". We want the whole path, not just the filename (agreed that the function name is misleading then. I didn't change it, just moved from `client/gpgcheck.c`).

Reply:
Okay, I didn't see that this was a move. Perhaps it should stay within gpgcheck.c if it's not a generic utility function?

Reply:
It *is* a generic utility function. I renamed it to a more accurate name.